src/ir/auth_logic/universe_relation_insertion.h

```diff
@@ -0,0 +1,449 @@
+  }
+
+  // This generates a set of says assertions that populate all the
+  // universes for all the speakers given the type environment. Essentially,
+  // it loops through the list of constant principals, and then all the
+  // constants of all types, and then generates a SaysAssertion that
+  // adds the constant to the universe for each speaker.
+  static std::vector<SaysAssertion> GetUniverseDefiningFacts(
+      const TypeEnv& type_env) {
+    std::vector<Principal> constant_principals;
+    for (const auto& typing : type_env) {
+      if (typing.second.kind() == datalog::ArgumentType::Kind::kPrincipal) {
+        constant_principals.push_back(Principal(typing.first));
+      }
+    }
+
+    std::vector<SaysAssertion> universe_defining_says_assertions;
+    for (const auto& prin : constant_principals) {
+      for (const auto& typing : type_env) {
+        universe_defining_says_assertions.push_back(SaysAssertion(
+            prin,  // speaker
+            {Assertion(Fact({}, MakeUniverseMembershipFact(typing.first,
+                                                           typing.second)))}));
+      }
+    }
+    return universe_defining_says_assertions;
+  }
+
+  // This visitor populates a mapping (flat_hash_map) of constant (literal)
+  // names to their types by looking at the relations to which the literals
+  // are applied and declarations of those relations (including the types
+  // of the parameters where they are applied). It takes as input
+  // a mapping from relation names to relation declarations.
+  class TypeEnvironmentGenerationVisitor
+      : public AuthLogicAstTraversingVisitor<TypeEnvironmentGenerationVisitor> {
+   public:
+    // A map from the names of principal / predicate argument literals
+    // to typings.
+    using TypeEnv = absl::flat_hash_map<std::string, datalog::ArgumentType>;
+    // A map from predicate names to predicate declarations.
+    using DeclEnv =
+        absl::flat_hash_map<std::string, datalog::RelationDeclaration>;
+
+    TypeEnvironmentGenerationVisitor(
+        absl::flat_hash_map<std::string, datalog::RelationDeclaration> decl_env)
+        : decl_env_(decl_env) {}
+
+    const TypeEnv& type_env() const { return type_env_; }
+
+   private:
+    void AddTyping(std::string arg_name, datalog::ArgumentType arg_type) {
+      if (!IsNameConstant(arg_name)) {
+        // This pass is only meant to find typings for constants,
+        // so do nothing if this argument is not a constant.
+        return;
+      }
+      if (type_env_.find(arg_name) == type_env_.end()) {
+        // No typing for this constant was found yet,
+        // so add it.
+        type_env_.insert({arg_name, arg_type});
+      } else if (type_env_.find(arg_name)->second != arg_type) {
+        // If there is already a typing for this constant,
+        // but it is different from the one used here,
+        // throw a type error and abort.
+        LOG(FATAL) << "type error for constant: " << arg_name;
+      }
```

Comment:
This can be made more concise as follows:

```C++
auto insert_result = type_env.insert({arg_name, arg_type});
CHECK(insert_result.second) << "Type error for constant: " << insert_result.first->first;
```

Reply:
fixed.